File: golden-paths/h2-enhancement/microservice/skeleton/python/main.py
"""${{ values.serviceName }} - FastAPI Microservice."""

import os
import time
from fastapi import FastAPI, HTTPException
from pydantic import BaseModel
from typing import Optional
from uuid import uuid4

app = FastAPI(
    title="${{ values.serviceName }}",
    version="1.0.0",
)

START_TIME = time.time()
items: dict[str, dict] = {}


class ItemCreate(BaseModel):
    name: str
    description: Optional[str] = None


class ItemUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None


@app.get("/health")
def health():
    return {"status": "healthy", "service": "${{ values.serviceName }}"}


@app.get("/ready")
def ready():
    return {"status": "ready", "uptime_seconds": round(time.time() - START_TIME, 2)}


@app.get("/metrics")
def metrics():
    return {
        "service": "${{ values.serviceName }}",
        "uptime_seconds": round(time.time() - START_TIME, 2),
        "items_count": len(items),
    }


@app.get("/api/items")
def list_items():
    return list(items.values())


@app.get("/api/items/{item_id}")
def get_item(item_id: str):
    if item_id not in items:
        raise HTTPException(status_code=404, detail="Item not found")
    return items[item_id]


@app.post("/api/items", status_code=201)
def create_item(item: ItemCreate):
    item_id = str(uuid4())
    record = {"id": item_id, **item.model_dump()}
    items[item_id] = record
    return record


@app.put("/api/items/{item_id}")
def update_item(item_id: str, item: ItemUpdate):
    if item_id not in items:
        raise HTTPException(status_code=404, detail="Item not found")
    existing = items[item_id]
    update_data = item.model_dump(exclude_unset=True)
    existing.update(update_data)
    return existing


@app.delete("/api/items/{item_id}", status_code=204)
def delete_item(item_id: str):
    if item_id not in items:
        raise HTTPException(status_code=404, detail="Item not found")
    del items[item_id]


if __name__ == "__main__":
    import uvicorn

    port = int(os.getenv("PORT", "${{ values.httpPort }}"))
    uvicorn.run(app, host="0.0.0.0", port=port)

File: golden-paths/h2-enhancement/microservice/skeleton/rust/main.rs
use actix_web::{web, App, HttpServer, HttpResponse, middleware};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::Mutex;
use std::time::Instant;
use uuid::Uuid;

struct AppState {
    start_time: Instant,
    items: Mutex<HashMap<String, Item>>,
}

#[derive(Serialize, Deserialize, Clone)]
struct Item {
    id: String,
    name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    description: Option<String>,
}

#[derive(Deserialize)]
struct ItemRequest {
    name: Option<String>,
    description: Option<String>,
}

async fn health() -> HttpResponse {
    HttpResponse::Ok().json(serde_json::json!({
        "status": "healthy",
        "service": "${{ values.serviceName }}"
    }))
}

async fn ready(data: web::Data<AppState>) -> HttpResponse {
    let uptime = data.start_time.elapsed().as_secs();
    HttpResponse::Ok().json(serde_json::json!({
        "status": "ready",
        "uptime_seconds": uptime
    }))
}

async fn metrics(data: web::Data<AppState>) -> HttpResponse {
    let uptime = data.start_time.elapsed().as_secs();
    let count = data.items.lock().unwrap().len();
    HttpResponse::Ok().json(serde_json::json!({
        "service": "${{ values.serviceName }}",
        "uptime_seconds": uptime,
        "items_count": count
    }))
}

async fn list_items(data: web::Data<AppState>) -> HttpResponse {
    let items = data.items.lock().unwrap();
    let result: Vec<&Item> = items.values().collect();
    HttpResponse::Ok().json(result)
}

async fn get_item(data: web::Data<AppState>, path: web::Path<String>) -> HttpResponse {
    let id = path.into_inner();
    let items = data.items.lock().unwrap();
    match items.get(&id) {
        Some(item) => HttpResponse::Ok().json(item),
        None => HttpResponse::NotFound().json(serde_json::json!({"error": "Item not found"})),
    }
}

async fn create_item(data: web::Data<AppState>, body: web::Json<ItemRequest>) -> HttpResponse {
    let name = match &body.name {
        Some(n) => n.clone(),
        None => return HttpResponse::BadRequest().json(serde_json::json!({"error": "name is required"})),
    };
    let item = Item {
        id: Uuid::new_v4().to_string(),
        name,
        description: body.description.clone(),
    };
    let mut items = data.items.lock().unwrap();
    items.insert(item.id.clone(), item.clone());
    HttpResponse::Created().json(item)
}

async fn update_item(
    data: web::Data<AppState>,
    path: web::Path<String>,
    body: web::Json<ItemRequest>,
) -> HttpResponse {
    let id = path.into_inner();
    let mut items = data.items.lock().unwrap();
    match items.get_mut(&id) {
        Some(item) => {
            if let Some(name) = &body.name {
                item.name = name.clone();
            }
            if let Some(desc) = &body.description {
                item.description = Some(desc.clone());
            }
            HttpResponse::Ok().json(item.clone())
        }
        None => HttpResponse::NotFound().json(serde_json::json!({"error": "Item not found"})),
    }
}

async fn delete_item(data: web::Data<AppState>, path: web::Path<String>) -> HttpResponse {
    let id = path.into_inner();
    let mut items = data.items.lock().unwrap();
    match items.remove(&id) {
        Some(_) => HttpResponse::NoContent().finish(),
        None => HttpResponse::NotFound().json(serde_json::json!({"error": "Item not found"})),
    }
}

#[actix_web::main]
async fn main() -> std::io::Result<()> {
    let port: u16 = std::env::var("PORT")
        .unwrap_or_else(|_| "${{ values.httpPort }}".to_string())
        .parse()
        .expect("PORT must be a number");

    let data = web::Data::new(AppState {
        start_time: Instant::now(),
        items: Mutex::new(HashMap::new()),
    });

    println!("${{ values.serviceName }} listening on port {}", port);

    HttpServer::new(move || {
        App::new()
            .app_data(data.clone())
            .route("/health", web::get().to(health))
            .route("/ready", web::get().to(ready))
            .route("/metrics", web::get().to(metrics))
            .route("/api/items", web::get().to(list_items))
            .route("/api/items/{id}", web::get().to(get_item))
            .route("/api/items", web::post().to(create_item))
            .route("/api/items/{id}", web::put().to(update_item))
            .route("/api/items/{id}", web::delete().to(delete_item))
    })
    .bind(("0.0.0.0", port))?
    .run()
    .await
}
